src/mvs/patch_match_net.cc

```diff
@@ -0,0 +1,156 @@
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Author: Antonios Matakos (anmatako-at-microsoft-dot-com)
+
+#include "mvs/patch_match_net.h"
+#include "util/misc.h"
+#include "torch/script.h"
+
+namespace colmap {
+namespace mvs {
+
+static const torch::Device kDevOut(torch::kCPU);
+static const torch::Device kDevIn(torch::cuda::is_available() ? torch::kCUDA
+                                                              : torch::kCPU);
+
+std::unordered_map<int, torch::jit::Module> PatchMatchNet::model_;
+
+PatchMatchNet::PatchMatchNet(const PatchMatchOptions& options,
+                             const PatchMatch::Problem& problem,
+                             const int thread_index)
+    : PatchMatch(options, problem), thread_index_(thread_index) {
+  InitModule();
+  InitProblemInputs();
+}
+
+void PatchMatchNet::Run() {
+  // Gradients should be disabled during evaluation to save up on GPU memory
+  torch::NoGradGuard no_grad;
+  std::cout << "Starting PatchMatchNet..." << std::endl;
+
+  // Run TorchScript module
+  torch::IValue result = model_[thread_index_].forward(
+      {images_, intrinsics_, extrinsics_, depth_params_});
+
+  // Copy tensor data to outputs
+  size_t width = problem_.images->at(problem_.ref_image_idx).GetWidth();
+  size_t height = problem_.images->at(problem_.ref_image_idx).GetHeight();
+  torch::Tensor depth = result.toTuple()->elements()[0].toTensor().to(kDevOut);
+  const float* depth_ptr = depth.data_ptr<float>();
+  std::copy(depth_ptr, depth_ptr + (width * height), depth_map_.GetPtr());
+
+  torch::Tensor confidence =
+      result.toTuple()->elements()[1].toTensor().to(kDevOut);
+  const float* confidence_ptr = confidence.data_ptr<float>();
+  std::copy(confidence_ptr, confidence_ptr + (width * height),
+            confidence_map_.GetPtr());
+}
+
+void PatchMatchNet::InitModule() {
+  torch::jit::setGraphExecutorOptimize(true);
+  // Load module only once per thread index from the execution thread-pool.
+  // Ensures we use execution optimization in a thread-safe manner
+  if (model_.count(thread_index_) == 0) {
+    std::cout << "First definition of patch-match module for thread index: "
+              << options_.gpu_index << std::endl;
+    model_[thread_index_] =
+        torch::jit::load(options_.mvs_module_path, kDevIn);
+  } else {
+    std::cout << "Patch-match module already defined for thread index: "
+              << options_.gpu_index << std::endl;
+  }
```

Comment:
Only the first run is successful, otherwise the thread terminates at model.forwards(...) (https://github.com/colmap/colmap/pull/1129/files#diff-ec9150c5522870ad0fd07f523905377b4d0670e7f42d92f2bbe11ceb42adb1beR59).

When I load the pytorch module every time this problem doesn't occur

Reply:
That's a very interesting failure mode. Failing at the `forward()` evaluation likely means that the module is not there for the subsequent runs. Are you executing on a single or multi GPU environment? If running on multi-GPU environment, try using just a single GPU index in the patch-match options and see if it fails the same way.

The reason I ended up with this setup instead of loading the module for each problem is to take advantage of the optimizations that LibTorch does in the JIT modules. My main finding was that loading the module every time and no optimization as about 2x slower compared to reusing the module and allow optimizations.

If you can share a dataset that causes this failure I can try to reproduce on my end and see if I can debug it effectively.

Reply:
I'm using a single GPU, not loading the model each time makes sense.

I'll send you the dataset.